### PR TITLE
copy merge fix #814

### LINKED ISSUE #814: copyMerge doesn't scale past 100k partitions
globAndSort calls standard string comparison.  This will produce incorrectly ordered files when nPartitions >= 100k

```
 in copyMerge:
    val srcFileStatuses = srcFilenames.flatMap(f => globAndSort(f))

  def globAndSort(filename: String): Array[FileStatus] = {
    val fs = fileSystem(filename)
    val path = new hadoop.fs.Path(filename)

    val files = fs.globStatus(path)
    if (files == null)
      return Array.empty[FileStatus]

    files.sortWith(_.compareTo(_) < 0)
  }
```

### PR DESCRIPTION
**Pull request by jigold**
@tpoterba @cseed Here's a different way of trying to fix this with partially applied functions. I think this is pretty messy, but fewer changes.

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala [hunk 1]
```diff
@@ -89,15 +89,18 @@ class RichHadoopConfiguration(val hConf: hadoop.conf.Configuration) extends AnyV
       }.toArray
   }
 
-  def globAndSort(filename: String): Array[FileStatus] = {
+  def globAndSort(filename: String, sortFn: (FileStatus) => Int = null): Array[FileStatus] = {
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala]: Should be `Option[(FileStatus) => Int]`

File: src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala [hunk 2]
```diff
@@ -107,7 +110,7 @@ class RichHadoopConfiguration(val hConf: hadoop.conf.Configuration) extends AnyV
       false, hConf)
   }
 
-  def copyMerge(srcFilenames: Array[String], destFilename: String, deleteSource: Boolean = true) {
+  def copyMerge(srcFilenames: Array[(String, (FileStatus) => Int)], destFilename: String, deleteSource: Boolean = true) {
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala]: No need to duplicate f: should be `copyMerge(srcFiles: Array[String], sortFunction: Option[(FileStatus) => Int] = None, ...)`

Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala]: Ah, I see why you did it this way -- the header.  Could also take a `headerFile: Option[String] = None` argument and put that first.

File: src/main/scala/org/broadinstitute/hail/utils/richUtils/RichRDDByteArray.scala [hunk 3]
```diff
@@ -10,6 +11,16 @@ import scala.reflect.ClassTag
 
 class RichRDDByteArray(val r: RDD[Array[Byte]]) extends AnyVal {
   def saveFromByteArrays(filename: String, header: Option[Array[Byte]] = None, deleteTmpFiles: Boolean = true) {
+    def getPartNumber(fileStatus: FileStatus): Int = {
+      val partRegex = """.*/?part-(\d+).*""".r
+
+      val fname = fileStatus.getPath.getName
+      fname match {
+        case partRegex(i) => i.toInt
+        case _ => throw new PathIOException(s"invalid parquet file `$fname'")
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/richUtils/RichRDDByteArray.scala]: not a parquet file here

File: src/main/scala/org/broadinstitute/hail/utils/package.scala [hunk 4]
```diff
@@ -210,6 +211,19 @@ package object utils extends Logging
 
   def uriPath(uri: String): String = new URI(uri).getPath
 
+  def getPartNumber(fileStatus: FileStatus): Int = getPartNumber(fileStatus.getPath.getName)
+
+  def getPartNumber(fname: String): Int = {
```
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/utils/package.scala]: prefer to split this into `getParquetPartNumber` and `getPartNumber`

File: src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala (hunk 2 above)
Comment (jigold) [inline src/main/scala/org/broadinstitute/hail/utils/richUtils/RichHadoopConfiguration.scala]: I think it's better to have the caller to copyMerge say explicitly how to sort the files. If you absolutely hate it, I can try and think about how to incorporate the header separately in copyMerge.

**Pull request closed without merging**